### PR TITLE
Add AMQP retry mechanism and example projects

### LINKED ISSUE #288: Amqp Connector don't work with RestartSource
HI guys!
I am trying to create a Reconnectable Source with RestartSource using Akka Streams Amqp.V1 and some strange things happen. When I disconnect the network, RestartSource does not try to reconnect again and no message is printed in the logs. However, when I receive an invalid message (type incompatible) it restarts my source. I am not sure why this is so and I would like to know if you have already been there and if you could help me. Below is the Source creation code:

```
using System.Numerics;
using Akka.Serialization;
using Akka.Streams.Amqp.V1.Dsl;
using Akka.Streams.Dsl;
using Amqp;
using System;
using Akka.Actor;
using Address = Amqp.Address;
using Akka.Streams;
using Akka.Streams.Amqp.V1;
using System.Text;

namespace image_processor_akkanet
{
    class Program
    {
        private static Connection connection;
        private static Session session;

        static void Main(string[] args)
        {
            var sys = ActorSystem.Create("AMQP-System");
            var materializer = ActorMaterializer.Create(sys);
            var serialization = sys.Serialization;
            var serializer = serialization.FindSerializerForType(typeof(byte[]));

            try
            {
                var address = new Address("0.0.0.0", 61616, "admin", "admin", scheme: "AMQP");
                connection = new Connection(address);
                session = new Session(connection);

                var queueName = "akka.teste";
                var receiverlinkName = "amqp-conn-test-sender";

                //create source
                var amqpSource = RestartSource.OnFailuresWithBackoff(
                    () => {
                        Console.WriteLine("Start/Restart...");
                        return AmqpSource
                          .Create(new NamedQueueSourceSettings<byte[]>(session, receiverlinkName, queueName, 200, serializer));
                    },
                    TimeSpan.FromSeconds(1),
                    TimeSpan.FromSeconds(3),
                    0.2,
                    Int16.MaxValue
                );

                //run source
                var result = amqpSource
                    .Throttle(1, TimeSpan.FromSeconds(1), 10, ThrottleMode.Shaping)
                    .Select(elem => Encoding.UTF8.GetString(elem, 0, elem.Length))
                    .RunForeach(elem => Console.WriteLine(elem), materializer);

                result.Wait(TimeSpan.FromSeconds(Int16.MaxValue));
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error " + ex.Message);
            }
            finally
            {
                session.Close();
                connection.Close();
            }
        }
    }
}
```
I tried using both **RestartSource.OnFailuresWithBackoff** and **RestartSource.WithBackoff** both to no avail

My project and dependencies:
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <RootNamespace>image_processor_akkanet</RootNamespace>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Akka" Version="1.4.11" />
    <PackageReference Include="Akka.Streams" Version="1.4.11" />
    <PackageReference Include="Akka.Streams.Amqp.V1" Version="1.0.0-beta2" />
  </ItemGroup>

</Project>
```

I'm using ActiveMQ Artemis as a AMQP Broker.

### PR DESCRIPTION
**Pull request by Arkatufus**
Closes #288

Adds settings variant that's able to manage the `Connection` and `Session` states so that it plays nice with disconnection/reconnection; all the user need to do is provide the `Address` object.

### CODE REVIEW

File: src/Amqp/Akka.Streams.Amqp.V1/AmqpSinkStage.cs [hunk 1]
```diff
@@ -49,15 +65,52 @@ public AmqpSinkStageLogic(AmqpSinkStage<T> amqpSinkStage, TaskCompletionSource<D
...
+                    catch (Exception e)
+                    {
+                        if (!_stage.AmqpSourceSettings.ManageConnection)
+                        {
+                            throw new ConnectionException(
+                                "Failed to connect to AMQP.V1 server. Could not retry connection because SinkSettings does not manage the Connection object.", e);
+                        }
+
+                        retry--;
+                        if (retry == 0)
```
Comment (Aaronontheweb) [inline src/Amqp/Akka.Streams.Amqp.V1/AmqpSinkStage.cs:90]: LGTM

File: src/Amqp/Akka.Streams.Amqp.V1/AmqpSourceStage.cs [hunk 2]
```diff
@@ -51,10 +69,45 @@ public override void PreStart()
             {
                 base.PreStart();
 
-                var consumerCallback = GetAsyncCallback<Message>(HandleDelivery);
-                _receiver.Start(_amqpSourceSettings.Credit, (_, m) => consumerCallback.Invoke(m));
+                Connect().Wait();
```
Comment (Aaronontheweb) [inline src/Amqp/Akka.Streams.Amqp.V1/AmqpSourceStage.cs:72]: Connect call is blocking here but there isn't much you can do about that - have to wait for the execution to finish somewhere in order to start the message pump.

Comment (Aaronontheweb) [reply src/Amqp/Akka.Streams.Amqp.V1/AmqpSourceStage.cs:72]: Actually, come to think of it - might want to just have the async state machine invoke a callback on the stage and signal failure / success. Given that this retry + backoff mechanism can run for several minutes, you don't want to tie the process up for that long. How hard would it be to do that?

File: src/Amqp/Akka.Streams.Amqp.V1/AmqpSinkStage.cs [hunk 3]
```diff
@@ -49,15 +65,52 @@ public AmqpSinkStageLogic(AmqpSinkStage<T> amqpSinkStage, TaskCompletionSource<D
...
+            }
+
+            private void HandleDisconnection((IAmqpObject sender, Error error) args)
+                => FailStage(new DisconnectedException(args.sender, args.error));
+
             public override void PreStart()
             {
                 base.PreStart();
+
+                Connect().Wait();
```
Comment (Aaronontheweb) [inline src/Amqp/Akka.Streams.Amqp.V1/AmqpSinkStage.cs:107]: Same comment as below about blocking on a really long-running call

File: src/Amqp/Examples/Amqp.V1.Sink/Program.cs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+            await _system.Terminate();
+        }
+
+        // The settings object manages the Session and Connection object
+        private static async Task UsingAddressSink()
+        {
+            var settings = new AddressSinkSettings<string>(_address, SenderLinkName, QueueName, _serializer);
+
+            // create sink
+            var amqpSink = RestartSink.WithBackoff(() =>
```
Comment (Aaronontheweb) [inline src/Amqp/Examples/Amqp.V1.Sink/Program.cs:43]: looks great

File: src/Amqp/Examples/Amqp.V1.Source/Amqp.V1.Source.csproj [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
```
Comment (Aaronontheweb) [inline src/Amqp/Examples/Amqp.V1.Source/Amqp.V1.Source.csproj:5]: Need to add `<IsPackable>false</IsPackable>` here or we're going to accidentally push a NuGet package. Dumb, I know, but I've done it before

File: src/Amqp/Examples/Amqp.V1.Sink/Amqp.V1.Sink.csproj [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
```
Comment (Aaronontheweb) [inline src/Amqp/Examples/Amqp.V1.Sink/Amqp.V1.Sink.csproj:5]: Need to add <IsPackable>false</IsPackable> here or we're going to accidentally push a NuGet package. Dumb, I know, but I've done it before

**Review by Aaronontheweb — CHANGES_REQUESTED** [review]
Need to find a way to `Connect` asynchronously without blocking - it's a bigger issue here due to the backoff.

File: src/Amqp/Akka.Streams.Amqp.V1/AmqpSourceStage.cs (hunk 2 above)
Comment (Arkatufus) [reply src/Amqp/Akka.Streams.Amqp.V1/AmqpSourceStage.cs:72]: I'll give it a look see, it shouldn't be that hard

File: src/Amqp/Examples/Amqp.V1.Source/Amqp.V1.Source.csproj (hunk 5 above)
Comment (Arkatufus) [reply src/Amqp/Examples/Amqp.V1.Source/Amqp.V1.Source.csproj:5]: Done

File: src/Amqp/Examples/Amqp.V1.Sink/Amqp.V1.Sink.csproj (hunk 6 above)
Comment (Arkatufus) [reply src/Amqp/Examples/Amqp.V1.Sink/Amqp.V1.Sink.csproj:5]: Done

File: src/Amqp/Akka.Streams.Amqp.V1/AmqpSourceStage.cs (hunk 2 above)
Comment (Arkatufus) [reply src/Amqp/Akka.Streams.Amqp.V1/AmqpSourceStage.cs:72]: On second thought, can we separate this into a different PR?
It "works" as it is, I'll work on the thread blocking issue on another PR.

**Review by Aaronontheweb — APPROVED** [review]

File: src/Amqp/Akka.Streams.Amqp.V1/AmqpSourceStage.cs (hunk 2 above)
Comment (Aaronontheweb) [reply src/Amqp/Akka.Streams.Amqp.V1/AmqpSourceStage.cs:72]: That's fine with me

**Pull request merged**